File: src/NodeParser/ConstructorNodeParser.ts
import ts from "typescript";
import { SubNodeParser } from "../SubNodeParser.js";
import { BaseType } from "../Type/BaseType.js";
import { ConstructorType } from "../Type/ConstructorType.js";
import { FunctionOptions } from "../Config.js";
import { NeverType } from "../Type/NeverType.js";
import { Context, NodeParser } from "../NodeParser.js";
import { DefinitionType } from "../Type/DefinitionType.js";
import { getNamedArguments, getTypeName } from "./FunctionNodeParser.js";

export class ConstructorNodeParser implements SubNodeParser {
    constructor(
        protected childNodeParser: NodeParser,
        protected functions: FunctionOptions,
    ) {}

    public supportsNode(node: ts.TypeNode): boolean {
        return node.kind === ts.SyntaxKind.ConstructorType;
    }

    public createType(node: ts.ConstructorTypeNode, context: Context): BaseType {
        if (this.functions === "hide") {
            return new NeverType();
        }

        const name = getTypeName(node);
        const func = new ConstructorType(node, getNamedArguments(this.childNodeParser, node, context));

        return name ? new DefinitionType(name, func) : func;
    }
}

File: internal/parser/constructor_node_parser.go
package parser

// Port of src/NodeParser/ConstructorNodeParser.ts.

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/config"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

type ConstructorNodeParser struct {
	childNodeParser NodeParser
	functions       config.FunctionOptions
}

func NewConstructorNodeParser(childNodeParser NodeParser, functions config.FunctionOptions) *ConstructorNodeParser {
	return &ConstructorNodeParser{childNodeParser: childNodeParser, functions: functions}
}

func (p *ConstructorNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindConstructorType
}

func (p *ConstructorNodeParser) CreateType(node *ast.Node, context *Context, _ *types.ReferenceType) types.Type {
	if p.functions == config.FunctionsHide {
		return &types.NeverType{}
	}

	name := GetTypeName(node)
	function := NewConstructorType(node, GetNamedArguments(p.childNodeParser, node, context))

	if name != "" {
		return types.NewDefinitionType(name, function)
	}
	return function
}
